ykrt/src/mt.rs

```diff
@@ -219,215 +217,206 @@ impl MTThread {
+            debug_assert!(!ls.is_parked());
+
+            let count = ls.count();
+            if count < self.inner.hot_threshold {
+                // Try incrementing this location's hot count. We make no guarantees that this will
+                // succeed because under heavy contention we can end up racing with many other
+                // threads and it's not worth our time to halt execution merely to have an accurate
+                // hot count. However, we do try semi-hard to enforce monotonicity (i.e. preventing
+                // the hot count from going backwards) which can happen if an "older" thread has
+                // been paused for a long time. Even in that case, though, we do not try endlessly.
+                let mut spinwait = SpinWait::new();
                 loop {
-                    let count = lp.number_data();
-                    if count >= self.inner.hot_threshold {
-                        if self.inner.tracer.is_some() {
-                            // This thread is already tracing another Location, so either another
-                            // thread needs to trace this Location or this thread needs to wait
-                            // until the current round of tracing has completed. Either way,
-                            // there's no point incrementing the hot count.
-                            return None;
-                        }
-                        let new_state = State::phase_tracing(Arc::clone(&self.inner.tid));
-                        if loc.compare_and_swap(lp, new_state, Ordering::Relaxed) == lp {
-                            Rc::get_mut(&mut self.inner).unwrap().tracer = Some((
-                                start_tracing(self.inner.tracing_kind),
-                                Arc::clone(&self.inner.tid),
-                            ));
-                            return None;
-                        }
-                        // We raced with another thread, which probably moved the Location from
-                        // PHASE_COUNTING to PHASE_TRACING.
-                        unsafe {
-                            Arc::from_raw(new_state.pointer_data::<u8>() as *mut u8);
-                        }
-                        return None;
-                    } else {
-                        let new_state = State::phase_counting(count + 1);
-                        let old_lp = lp;
-                        lp = loc.compare_and_swap(lp, new_state, Ordering::Relaxed);
-                        if lp == old_lp {
-                            return None;
-                        }
-                        // We raced with another thread.
-                        if lp.phase() != PHASE_COUNTING {
-                            // The other thread probably moved the Location from PHASE_COUNTING to
-                            // PHASE_TRACING.
-                            return None;
+                    match loc.compare_exchange_weak(
```

Comment:
I was about to suggest a simplification involving `fetch_add`, but this could go awry. If another thread transitions us out of counting, we'd end up incrementing an address!

Reply:
Here's a weird (and flawed) idea (that you may have already considered).

If we split the counting and address fields into separate atomic integers, then there's no harm in incrementing the count when we've already transitioned to tracing and then this block of code becomes much simpler.

This means the state no longer fits in a usize, which is unacceptable, but I wanted to think out of the box anyway...

This comment is unactionable :P

Reply:
Good point! I had vaguely thought about `fetch_add` but decided against it for no deep reason. I think you're right that it would be a bad idea!